yarn/src/main/scala/org/apache/spark/deploy/yarn/Client.scala

```diff
@@ -540,6 +561,29 @@ private[spark] class Client(
     amContainer
   }
 
+  def setupCredentials(): Unit = {
+    Option(args.principal) match {
+      case Some(principal) =>
+        Option(args.keytab) match {
+          case Some(keytabPath) =>
+            // Generate a file name that can be used for the keytab file, that does not conflict
+            // with any user file.
+            logInfo("Attempting to login to the Kerberos" +
+              s" using principal: $principal and keytab: $keytabPath")
+            val f = new File(keytabPath)
+            keytabFileName = f.getName + "-" + System.currentTimeMillis()
+            val ugi = UserGroupInformation.loginUserFromKeytabAndReturnUGI(principal, keytabPath)
+            credentials = ugi.getCredentials
+            loginFromKeytab = true
+            logInfo("Successfully logged into Kerberos.")
```

Comment:
I think with this change, the user would not need to kinit, since this method is used before any YARN or HDFS specific actions are made. So we do the login here and use those credentials.